### ISSUE TITLE
Improving correctness of gamma application: It shouldn't be applied to material color parameters

### ISSUE THREAD

**Issue by bhouston** [mrdoob/three.js#5838]
Right now when gammInput is turned on, gamma is applied to all input colors to the shader -- this is usually not how it is done and is generally not correct.

In particular, I am referring to these applications of input gamma: [diffuse color](https://github.com/mrdoob/three.js/blob/master/src/renderers/WebGLRenderer.js#L4617), [emissive color](https://github.com/mrdoob/three.js/blob/master/src/renderers/WebGLRenderer.js#L4754), [specular](https://github.com/mrdoob/three.js/blob/master/src/renderers/WebGLRenderer.js#L4755) as well as [vertex colors](https://github.com/mrdoob/three.js/blob/master/src/renderers/shaders/ShaderChunk/color_vertex.glsl#L5).  Also the [ambient light color](https://github.com/mrdoob/three.js/blob/master/src/renderers/WebGLRenderer.js#L5328), the [directional light color](https://github.com/mrdoob/three.js/blob/master/src/renderers/WebGLRenderer.js#L5359), the [point light color](https://github.com/mrdoob/three.js/blob/master/src/renderers/WebGLRenderer.js#L5379), the [spot light color](https://github.com/mrdoob/three.js/blob/master/src/renderers/WebGLRenderer.js#L5407), the [hemisphere light colors](https://github.com/mrdoob/three.js/blob/master/src/renderers/WebGLRenderer.js#L5458).

As @WestLangley likely knows, use of gamma is to compensate for the sRGB standard used in a number of files formats, particularly JPG, PNG, GIF, etc.  These images are pre-gamma corrected (by approximately pow( RGB, 1/2.2) so that when displayed on a standard monitor (which has a built in gamma of pow(RGB, 2.2) they look correct.

As we want to do rendering in linear space, we want bring the sRGB images into linear space, so we gamma correct them to remove their built-in gamma.

But all shader parameters are not pre-gamma corrected by pow( RGB, 1/ 2.2), they are linear, true values.  Thus to apply an input gamma correction, e.g. pow( RGB, 2.2 ), to them in incorrect, we are making our linear colors darker than they should be in the mid range, around 0.5.  This gets progressively worse as you get HDR colors, such as light colors above (1,1,1) (or Light.intensity), as applying the gamma input "correction" will make them quadratically brighter than they should be.

So I am proposing removing gamma correction from all inputs that are not texture inputs.  This will simplify the code, make it smaller and make it more correct.

**Comment by bhouston** [comment]
@WestLangley BTW: https://www.khronos.org/registry/webgl/extensions/EXT_sRGB/

**Comment by mrdoob** [comment]
This all sounds great to me.
Thanks a lot for the explanations by the way, I never had a good understanding of all this O:)

**Comment by Wandalen** [comment]
Hi guys. Currently gamma factor used by:

- Materials
- Lights
- Renderer

I would say it should be in Material exclusively. That's my suggestion.

**Comment by Wandalen** [comment]
Most challenging part to move it out of lights.
To achieve it it is possible to make several color uniform a light type.
For example for spot light:
```
  'spotLightColor' : new Uniform({ type: 'fv', value: [] }),
  'spotLightColorGamma' : new Uniform({ type: 'fv', value: [] }),
```
And in shaders we will get:
```
    '#ifdef GAMMA_INPUT',
    '  uniform vec3 ambientLightColorGamma;',
    '  #define ambientLightColor ambientLightColorGamma';
    '#else'
    '  uniform vec3 ambientLightColor;',
    '#endif GAMMA_INPUT',

    '#if MAX_DIR_LIGHTS > 0',

    '  #ifdef GAMMA_INPUT',
    '    uniform vec3 directionalLightColorGamma[ MAX_DIR_LIGHTS ];',
    '    #define directionalLightColor directionalLightColorGamma';
    '  #else'
    '    uniform vec3 directionalLightColor[ MAX_DIR_LIGHTS ];',
    '  #endif GAMMA_INPUT',

    '  uniform vec3 directionalLightDirection[ MAX_DIR_LIGHTS ];',

    '#endif',

    '#if MAX_HEMI_LIGHTS > 0',

    '  #ifdef GAMMA_INPUT',
    '    uniform vec3 hemisphereLightSkyColorGamma[ MAX_DIR_LIGHTS ];',
    '    #define hemisphereLightSkyColor hemisphereLightSkyColorGamma';
    '  #else'
    '    uniform vec3 hemisphereLightSkyColor[ MAX_DIR_LIGHTS ];',
    '  #endif GAMMA_INPUT',

    '  #ifdef GAMMA_INPUT',
    '    uniform vec3 hemisphereLightGroundColorGamma[ MAX_DIR_LIGHTS ];',
    '    #define hemisphereLightGroundColor hemisphereLightGroundColorGamma';
    '  #else'
    '    uniform vec3 hemisphereLightGroundColor[ MAX_DIR_LIGHTS ];',
    '  #endif GAMMA_INPUT',

    '  uniform vec3 hemisphereLightDirection[ MAX_HEMI_LIGHTS ];',

    '#endif',

    '#if MAX_POINT_LIGHTS > 0',

    '  #ifdef GAMMA_INPUT',
    '    uniform vec3 pointLightColorGamma[ MAX_DIR_LIGHTS ];',
    '    #define pointLightColor pointLightColorGamma';
    '  #else'
    '    uniform vec3 pointLightColor[ MAX_DIR_LIGHTS ];',
    '  #endif GAMMA_INPUT',

    '  uniform vec3 pointLightPosition[ MAX_POINT_LIGHTS ];',
    '  uniform float pointLightDistance[ MAX_POINT_LIGHTS ];',

    '#endif',

    '#if MAX_SPOT_LIGHTS > 0',

    '  #ifdef GAMMA_INPUT',
    '    uniform vec3 spotLightColorGamma[ MAX_DIR_LIGHTS ];',
    '    #define spotLightColor spotLightColorGamma';
    '  #else'
    '    uniform vec3 spotLightColor[ MAX_DIR_LIGHTS ];',
    '  #endif GAMMA_INPUT',

    '  uniform vec3 spotLightPosition[ MAX_SPOT_LIGHTS ];',
    '  uniform vec3 spotLightDirection[ MAX_SPOT_LIGHTS ];',
    '  uniform float spotLightAngleCos[ MAX_SPOT_LIGHTS ];',
    '  uniform float spotLightExponent[ MAX_SPOT_LIGHTS ];',

    '  uniform float spotLightDistance[ MAX_SPOT_LIGHTS ];',

    '#endif',
```

**Comment by Wandalen** [comment]
It's easier to make it parameter material-specific. Later it could be done texture-specific. But texture is not only thing which depends of Gamma.

**Comment by vorg** [comment]
How is that supposed to work in practice?

If I don't correct input but correct the output then the color with be brighter, no?

Let's say I have material

```javascript
var m = new THREE.MeshLambertMaterial( { color: 0xe74c3c, shading: THREE.FlatShading } ) ;
```

I'm trying to use peach #74c3c color:

<img width="224" alt="screen shot 2015-09-18 at 02 05 30" src="https://cloud.githubusercontent.com/assets/171001/9949787/c94bd2fc-5da9-11e5-81f0-22a9b39cb6d3.png">


![screen shot 2015-09-18 at 02 00 49](https://cloud.githubusercontent.com/assets/171001/9949771/7e8128d0-5da9-11e5-82e6-ec6f23852946.jpg)

- On the left is Unity 5 with linear space lighting
- In the middle default ThreeJS setup  (`gammaInput` and `gammaOutput` set to false) with valid color but invalid light falloff (gradient)
- On the rightThreeJS setup (`gammaInput` and `gammaOutput` set to true) sharp light falloff due to gamma correction but invalid (too bright) color

**Comment by bhouston** [comment]
Solved via: PR #8117
